### PR TITLE
match lowering: Hide `Candidate` from outside the lowering algorithm

### PR DESCRIPTION
**Pull request by Nadrieril**
The internals of `Candidate` are tricky and a source of confusion. This PR makes it so we don't expose `Candidate`s outside the lowering algorithm. Now:
- false edges are handled in `lower_match_tree`;
- `lower_match_tree` takes a list of patterns as input;
- `lower_match_tree` returns a flat datastructure that contains only the necessary information.

r? @matthewjasper

try-job: x86_64-rust-for-linux

### CODE REVIEW

File: compiler/rustc_mir_build/src/build/matches/mod.rs [hunk 1]
```diff
@@ -1305,32 +1378,83 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
...
+                next_candidate_start_block = leaf_candidate.false_edge_start_block;
             });
         }
 
-        if refutable {
-            // In refutable cases there's always at least one candidate, and we want a false edge to
-            // the failure block.
-            previous_candidate.as_mut().unwrap().next_candidate_start_block = Some(otherwise_block)
-        } else {
+        if !refutable {
```
Comment (Zalathar) [inline compiler/rustc_mir_build/src/build/matches/mod.rs:1457]: Side note: `refutable: bool` seems like another good candidate for a dedicated enum in the future.

Comment (Nadrieril) [reply compiler/rustc_mir_build/src/build/matches/mod.rs:1457]: Possibly yes. I have hopes to remove the need for it, we'll see

File: compiler/rustc_mir_build/src/build/matches/mod.rs [hunk 2]
```diff
@@ -1305,32 +1378,83 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
...
+            scrutinee_place_builder.base(),
+        );
+
         // See the doc comment on `match_candidates` for why we have an otherwise block.
         let otherwise_block = self.cfg.start_new_block();
 
         // This will generate code to test scrutinee_place and branch to the appropriate arm block
-        self.match_candidates(match_start_span, scrutinee_span, block, otherwise_block, candidates);
+        let mut candidate_refs =
+            candidates.iter_mut().map(|candidate| candidate).collect::<Vec<_>>();
```
Comment (Zalathar) [inline compiler/rustc_mir_build/src/build/matches/mod.rs:1410]: Nitpick: I tried removing this `map` and it still works, so the map seems to be unnecessary.

**Comment by Zalathar** [comment]
So if I understand this correctly:

- There is a flat list of `MatchTreeBranch` at the top level, and each one corresponds 1:1 with an arm of the match expression.
  - Should we assert that this is true just before zipping the arms/branches back together?
- Each of those branches contains a flat list of 1 or more `MatchTreeSubBranch`.
  - Should we assert that the list of sub-branches is not empty?
  - A sub-branch represents a “way” for that arm to decide that it should jump to checking the guard (or jump to the arm body if the arm has no guard).
  - Arms without or-patterns should have exactly 1 sub-branch.
  - For arms with or-patterns, the exact number of sub-branches is an implementation detail, depending on opportunities for simplification (which typically require there to be no guard).
  - The guard needs to be lowered separately for each sub-branch (because each sub-branch might bind different values).
  - If the guard fails, we jump to that sub-branch's `otherwise_block`, which will have been wired up to jump to the next sub-branch (if one exists), or otherwise to the next top-level arm branch.

**Comment by Nadrieril** [comment]
Yep, all correct. I don't feel the need to assert the number of arms, it's all preserved in `lower_match_tree`.

**Comment by Zalathar** [comment]
I was playing around with this, and one thing I found was that it's fairly easy to change `MatchTreeBranch::from_candidate` and `MatchTreeSubBranch::from_sub_candidate` so that they take `&Candidate` by reference instead of `Candidate` by value.

Obviously both ways are possible (since the candidates aren't used afterwards), but the by-reference version feels a bit nicer to me, because there's no need to worry about being able to move things at the right time.

(Not a blocking concern; I'm happy to propose this change as a follow-up PR.)

**Comment by Nadrieril** [comment]
> I was playing around with this, and one thing I found was that it's fairly easy to change `MatchTreeBranch::from_candidate` and `MatchTreeSubBranch::from_sub_candidate` so that they take `&Candidate` by reference instead of `Candidate` by value.
> 
> Obviously both ways are possible (since the candidates aren't used afterwards), but the by-reference version feels a bit nicer to me, because there's no need to worry about being able to move things at the right time.

I made them take `Candidate` exactly because it feels important that we don't use the `Candidate`s afterward. These functions are meant to be used exactly in one spot, why would there be a worry about being able to move things?

> Similarly, it's possible to have `lower_match_tree` take a `&BuiltMatchTree` by reference instead of `BuiltMatchTree` by value, with only minimal tweaks to the relevant code.

I'm guessing you mean `bind_pattern` could take `&MatchTreeBranch`? Sure, that one feels less important. But again I'm curious what your usecase is, given that this is meant to be called exactly once per branch directly after `lower_match_tree`.

**Comment by Zalathar** [comment]
Ah, I didn't have a specific use case (beyond trying to get rid of the slightly ugly length check and `try_into` in `bind_pattern`), so if you did it this way on purpose then I won't worry about it.

**Comment by tgross35** [comment]
I _think_ this may have somehow failed the rollup at https://github.com/rust-lang/rust/pull/128454#issuecomment-2261480246. I am not entirely sure how but this seems like the only PR in the rollup that affects codegen.

I'll start a try job to confirm. The RFL job failed but I assume that was just the first CI job to complete and others would have failed too. Also going to ask on Zulip.

@bors r-

**Comment by tgross35** [comment]
I am totally wrong, the wasm PR updated stdarch. Sorry about that, please re-r+ after the try job completes.

**Pull request merged**